src/fastgltf_tools.hpp

```diff
@@ -0,0 +1,486 @@
+							accessor.sparse->indexComponentType, indicesBytes + indexStride * sparseIndexCount);
+				}
+			} else if (accessor.bufferViewIndex) {
+				func(internal::getAccessorElementAt<ElementType>(accessor.componentType,
+						srcBytes + srcStride * i));
+			} else {
+				func(ElementType{});
+			}
+		}
+
+		return;
+	}
+
+	// 5.1.1. accessor.bufferView
+	// The index of the buffer view. When undefined, the accessor MUST be initialized with zeros; sparse
+	// property or extensions MAY override zeros with actual values.
+	if (!accessor.bufferViewIndex) {
+		for (std::size_t i = 0; i < accessor.count; ++i) {
+			func(ElementType{});
+		}
+	}
+	else {
+		auto& view = asset.bufferViews[*accessor.bufferViewIndex];
+		auto stride = view.byteStride ? *view.byteStride : getElementByteSize(accessor.type, accessor.componentType);
+
+		auto* bytes = adapter(asset.buffers[view.bufferIndex]);
+		bytes += view.byteOffset + accessor.byteOffset;
+
+		for (std::size_t i = 0; i < accessor.count; ++i) {
+			func(internal::getAccessorElementAt<ElementType>(accessor.componentType, bytes + i * stride));
+		}
+	}
+}
+
+template <typename ElementType, std::size_t TargetStride = sizeof(ElementType),
+		 typename BufferDataAdapter = DefaultBufferDataAdapter>
+#if FASTGLTF_HAS_CONCEPTS
+requires Element<ElementType>
+#endif
+void copyFromAccessor(const Asset& asset, const Accessor& accessor, void* dest,
+		const BufferDataAdapter& adapter = {}) {
+	using Traits = ElementTraits<ElementType>;
+	static_assert(Traits::type != AccessorType::Invalid, "Accessor traits must provide a valid Accessor Type");
+	static_assert(std::is_default_constructible_v<ElementType>, "Element type must be default constructible");
+	static_assert(std::is_constructible_v<ElementType>, "Element type must be constructible");
+	static_assert(std::is_move_assignable_v<ElementType>, "Element type must be move-assignable");
+
+	if (accessor.type != Traits::type) {
+		return;
+	}
+
+	auto* dstBytes = reinterpret_cast<std::byte*>(dest);
+
+	if (accessor.sparse && accessor.sparse->count > 0) {
+		std::size_t index = 0;
+		iterateAccessor<ElementType>(asset, accessor, [&](auto&& value) {
+			auto* pDest = reinterpret_cast<ElementType*>(dstBytes + TargetStride * index);
+			*pDest = std::forward<ElementType>(value);
+			++index;
+		}, adapter);
+
+		return;
+	}
+
+	auto elemSize = getElementByteSize(accessor.type, accessor.componentType);
+
+	// 5.1.1. accessor.bufferView
+	// The index of the buffer view. When undefined, the accessor MUST be initialized with zeros; sparse
+	// property or extensions MAY override zeros with actual values.
+	if (!accessor.bufferViewIndex) {
+		if constexpr (std::is_trivially_copyable_v<ElementType>) {
+			if (TargetStride == elemSize) {
+				std::memset(dest, 0, elemSize * accessor.count);
+			} else {
+				for (std::size_t i = 0; i < accessor.count; ++i) {
+					std::memset(dstBytes + i * TargetStride, 0, elemSize);
+				}
+			}
+		} else {
```

Comment:
`std::is_trivially_copyable_v` does not guarantee that a "default constructed" value is zero'd. It only guarantees that you can `std::memcpy` between a byte representation or other objects of the same type. I think the sensible option would be to check `std::is_trivially_constructible_v<ElementType, std::uint32_t>` (or std::byte?) to let them default construct with `0`: `ElementType { 0 }`. Though then you could also just use the branch below.

Or am I missing something here?

Reply:
This is a valid concern, however the highlighted code here is actually the correct part. using `std::memset` with trivially_copyable_v basically pretends that the source data is a bunch of zeroes and we are copying these zero bytes from them.

The problem arises in the other branch where we make the assumption that `ElementType{}` is the same as zero-initializing, but ultimately I think this is a safe assumption to make, and the alternate solutions might involve a significant amount of template trickery and an increased number of assumptions about the implementation of ElementType.
For example, ElementType{0} on a non-scalar type assumes that it A) has a single-arg constructor B) would correctly zero-fill if its first argument is 0. Another assumption we could make to fix that is to assume a ComponentCount-arg constructor, and come up with some templating to zero-fill the constructor with n zeroes, but that still makes the assumption that such a constructor exists and produces the correct result.

Reply:
After thinking about it some more, one possible solution is to simply request the user provide a default value, a la `std::vector.resize(size_type count, const value_type& value)`. This has the benefit of being more extensible by the user and makes the least assumptions about present constructors, but requires the user to be aware that some of their math types may need to have their zero-value be explicitly specified. This may also result in code duplication, since a 1-arg and 2-arg version would have to be implemented separately, as with vector resize, to minimize assumptions about copyability and constructability.

In all, I think this is a bit overkill considering the types specified are all numeric types, and it should be a safe assumption to make in general that the default constructor of a numeric type is the equivalent of zero in that number space.

Reply:
Yeah I guess it's better the way it is currently. I think this is fine as-is, and if anybody complains about needing some specific use-case we can see later. I guess we can leave it, and perhaps add some checks in the future to ensure that the types are numeric or can be constructed properly from "nothing".